### PR TITLE
SAIC-543: Cinder volume migration

### PR DESCRIPTION
**Pull request by aguzikova**

### CODE REVIEW

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.os.storage.cinder_database
cloudferrylib/os/storage/cinder_database.py:223: [W1202(logging-format-interpolation), CinderStorage.deploy_data_to_table.filter_data] Use % formatting in logging functions but pass the % parameters as arguments
************* Module cloudferrylib.os.storage.cinder_storage
cloudferrylib/os/storage/cinder_storage.py:107: [W0221(arguments-differ), CinderStorage.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/storage/cinder_storage.py:165: [W1201(logging-not-lazy), CinderStorage.finish] Specify string format arguments as logging function parameters
cloudferrylib/os/storage/cinder_storage.py:264: [W0613(unused-argument), CinderStorage.convert_snapshot] Unused argument 'cloud'
************* Module cloudferrylib.utils.remote_runner
cloudferrylib/utils/remote_runner.py:31: [W0621(redefined-outer-name), RemoteRunner.__init__] Redefining name 'sudo' from outer scope (line 15)
************* Module tests.cloudferrylib.os.storage.test_cinder_storage
tests/cloudferrylib/os/storage/test_cinder_storage.py:47: [E1002(super-on-old-class), CinderStorageTestCase.setUp] Use of super on an old style class
tests/cloudferrylib/os/storage/test_cinder_storage.py:233: [W0703(broad-except), CinderStorageTestCase.test_get_volume_path_iscsi_error] Catching too general exception Exception
************* Module tests.generate_load
devlab/tests/generate_load.py:7: [W0403(relative-import), ] Relative import 'config', should be 'tests.config'
devlab/tests/generate_load.py:9: [F0401(import-error), ] Unable to import 'filtering_utils'
devlab/tests/generate_load.py:36: [W0621(redefined-outer-name), clean_if_exists.wrapper] Redefining name 'args' from outer scope (line 1188)
devlab/tests/generate_load.py:971: [W0120(useless-else-on-loop), CleanEnv.clean_vms.wait_until_vms_all_deleted] Else clause on loop without a break statement
devlab/tests/generate_load.py:977: [W0106(expression-not-assigned), CleanEnv.clean_vms] Expression "[vms.append(vm) for vm in self.config.vms_from_volumes]" is assigned to nothing
devlab/tests/generate_load.py:1122: [W0141(bad-builtin), CleanEnv.clean_keypairs.delete_user_keypairs] Used builtin function 'map'

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,6 @@ snapshots
 build
 dist
 CloudFerry.egg-info
+.ropeproject
+.venv
```
Comment (SVilgelm) [inline .gitignore]: .venv and .ropeproject should be added into .global_ignore

File: cloudferrylib/os/actions/cinder_database_manipulation.py [hunk 2]
```diff
@@ -47,18 +110,314 @@ def run(self, *args, **kwargs):
...
+        res = dict(
+            [(vt['id'], self.dst_volume_types[vt['name']])
+             for vt in _volume_types(self.data)
+             if vt['name'] in self.dst_volume_types])
+        return res
+
+    def _dst_host(self, vtid=None):
+        # vtid -> dst_host
+        # cached property
+        if not hasattr(self, 'dst_hosts') or self.dst_hosts is None:
```
Comment (SVilgelm) [inline cloudferrylib/os/actions/cinder_database_manipulation.py]: dst_host is always initialized in the __init__ function, "self.dst_hosts is None" is enough.

File: cloudferrylib/os/actions/cinder_database_manipulation.py [hunk 3]
```diff
@@ -47,18 +110,314 @@ def run(self, *args, **kwargs):
...
+                self.dst_mount[vt['id']] = []
+                output = self._mount_output(self.cloud[DST], vt)
+                for line in output:
+                    self.dst_mount[vt['id']].append(line.split(MOUNT_DELIM))
+
+        return self.dst_mount
+
+    def _dir_to_provider(self, dst):
+        # cached property
+        if not hasattr(self, 'dst_dir_to_provider') \
```
Comment (SVilgelm) [inline cloudferrylib/os/actions/cinder_database_manipulation.py]: hasattr() is not needed here.

File: cloudferrylib/os/actions/cinder_database_manipulation.py [hunk 4]
```diff
@@ -47,18 +110,314 @@ def run(self, *args, **kwargs):
...
+                "awk '{if (match($1, \"^'$exp'$\") && $3 ~ \"cinder\") "
+                "{print $3}}'; done"
+            )
+        res = self._run_cmd(cloud, cmd)
+        return res if isinstance(res, list) else [res]
+
+    def _vt_map(self):
+        # host volume_type_id->hostname map
+        # cached property
+        if not hasattr(self, 'dst_volume_types') \
```
Comment (SVilgelm) [inline cloudferrylib/os/actions/cinder_database_manipulation.py]: hasattr() is not needed here.

File: cloudferrylib/os/actions/cinder_database_manipulation.py [hunk 5]
```diff
@@ -47,18 +110,314 @@ def run(self, *args, **kwargs):
...
+
+        for i in self.cloud:
+            for sd in sorted(paths[i][BY_VTID].values()):
+                paths[i][ALL].extend(sd)
+
+        return paths
+
+    def _paths(self, position, vtid=None):
+        # cached property
+        if not hasattr(self, 'path_map') or self.path_map is None:
```
Comment (SVilgelm) [inline cloudferrylib/os/actions/cinder_database_manipulation.py]: hasattr() is not needed here.

File: cloudferrylib/os/actions/cinder_database_manipulation.py [hunk 6]
```diff
@@ -47,18 +110,314 @@ def run(self, *args, **kwargs):
...
+    def _dst_host(self, vtid=None):
+        # vtid -> dst_host
+        # cached property
+        if not hasattr(self, 'dst_hosts') or self.dst_hosts is None:
+            self.dst_hosts = \
+                [i.host for i in
+                 self.cloud[DST][RES].cinder_client.services.list(
+                     binary=CINDER_VOLUME) if i.state == 'up']
+        # cached property
+        if not hasattr(self, 'dst_volume_types') \
```
Comment (SVilgelm) [inline cloudferrylib/os/actions/cinder_database_manipulation.py]: hasattr() is not needed here.

File: cloudferrylib/os/actions/cinder_database_manipulation.py [hunk 7]
```diff
@@ -47,18 +110,314 @@ def run(self, *args, **kwargs):
...
+                _, t = h.split('@')
+                if t in self.dst_volume_types:
+                    host_map[self.dst_volume_types[t]] = h
+
+        host = host_map.get(vtid, self.dst_hosts[0])
+        return host
+
+    def _dst_mount_info(self):
+        # cached property
+        if not hasattr(self, 'dst_mount') or self.dst_mount is None:
```
Comment (SVilgelm) [inline cloudferrylib/os/actions/cinder_database_manipulation.py]: hasattr() is not needed here.

File: cloudferrylib/os/actions/cinder_database_manipulation.py [hunk 8]
```diff
@@ -47,18 +110,314 @@ def run(self, *args, **kwargs):
...
+                          for t in mount_info[vt['id']]])
+            else:
+                self.dst_dir_to_provider = \
+                    dict([t for t in mount_info[DEFAULT]])
+
+        return self.dst_dir_to_provider[dst]
+
+    def _provider_to_vtid(self, provider):
+        # cached property
+        if not hasattr(self, 'dst_provider_to_vtid') \
```
Comment (SVilgelm) [inline cloudferrylib/os/actions/cinder_database_manipulation.py]: hasattr() is not needed here.

File: cloudferrylib/os/storage/cinder_database.py [hunk 9]
```diff
@@ -98,36 +148,35 @@ def list_of_dicts_for_table(self, table):
...
-        if kwargs.get('tenant_id'):
-            self.filter_tenant_id = kwargs['tenant_id'][0]
 
         return jsondate.dumps(
             {i: self.list_of_dicts_for_table(i) for i in self.list_of_tables})
 
     def get_volume_host(self):
         # cached property
-        if not hasattr(self, 'hosts'):
+        if not hasattr(self, 'hosts') or self.hosts is None:
```
Comment (SVilgelm) [inline cloudferrylib/os/storage/cinder_database.py]: hasattr() is not needed here.

File: cloudferrylib/os/storage/cinder_database.py [hunk 10]
```diff
@@ -98,36 +148,35 @@ def list_of_dicts_for_table(self, table):
...
             self.hosts = [i.host for i in self.cinder_client.services.list(
-                binary=CINDER_VOLUME)]
+                binary=CINDER_VOLUME) if i.state == 'up']
         # return host by "round-robin" rule
         if not hasattr(self, "host_counter"):
```
Comment (SVilgelm) [inline cloudferrylib/os/storage/cinder_database.py]: hasattr() is not needed here.

File: cloudferrylib/os/storage/cinder_storage.py [hunk 11]
```diff
@@ -60,6 +62,20 @@ def get_client(self, params=None):
...
+            service_type='volume',
+            endpoint_type='publicURL',
+        )
+
+    def get_endpoint_host(self):
+        return self.extract_host(self.get_endpoint())
+
+    @staticmethod
+    def extract_host(url):
+        m = re.search('//([^:^/]*)', url)
```
Comment (SVilgelm) [inline cloudferrylib/os/storage/cinder_storage.py]: I think, we can use a compiled object, like:

    RE_EXRACT_HOST = re.compile('//([^:^/]*)')
...

    m = RE_EXRACT_HOST.search(url)

File: .gitignore (hunk 1 above)
Comment (aguzikova) [inline .gitignore]: ok

File: cloudferrylib/os/storage/cinder_storage.py (hunk 11 above)
Comment (aguzikova) [inline cloudferrylib/os/storage/cinder_storage.py]: ok

File: cloudferrylib/os/actions/cinder_database_manipulation.py (hunk 8 above)
Comment (aguzikova) [inline cloudferrylib/os/actions/cinder_database_manipulation.py]: It was hasattr() only (no "is not None), but then pylint didn't like it. =)

**Pull request closed without merging**